frontend/src/scenes/sceneLogic.ts

```diff
@@ -296,6 +303,54 @@ export const sceneLogic = kea<sceneLogicType<Scene, Params, LoadedScene, SceneCo
             posthog.capture('$pageview')
             document.title = values.scene ? `${identifierToHuman(values.scene)} • PostHog` : 'PostHog'
         },
+        openScene: ({ scene, params }) => {
+            const sceneConfig = sceneConfigurations[scene] || {}
+            const { user } = userLogic.values
+            const { preflight } = preflightLogic.values
+
+            if (scene === Scene.Signup && preflight && !preflight.cloud && preflight.initiated) {
+                // If user is on an initiated self-hosted instance, redirect away from signup
+                router.actions.replace('/login')
+                return
+            }
+
+            if (user) {
+                // If user is already logged in, redirect away from unauthenticated-only routes like signup
+                if (sceneConfig.onlyUnauthenticated) {
+                    if (scene === Scene.Login && router.values.searchParams['next']) {
+                        router.actions.replace(router.values.searchParams['next'])
```

Comment:
https://github.com/PostHog/posthog/blob/2b4290959108dd6401449e9a7c1251e2dbcb980e/frontend/src/scenes/authentication/loginLogic.ts#L41 already handles this redirection, and further I think it might be better to do a hard reload here in case we don't properly handle reloading everything?

Reply:
This only applies if you just log in. If while logged in you end up on the URL https://app.posthog.com/login?next=/events (e.g. you reload the login page), there would have been no redirection. It's there now.

Reply:
That said, I [consolidated](https://github.com/PostHog/posthog/pull/4642/commits/65865fc8f5de1153150080d07eee7b81f38fe6d1) the logic and added `afterLoginRedirect(): string` that gives both logics the url to open after the user has logged in.